### ISSUE TITLE
Create a version of switchIfEmpty that takes a Supplier<? extends Mono<?>>

### ISSUE THREAD

**Issue by danapsimer** [reactor/reactor-core#1164]
### Expected behavior
There should be an operation that can supply an alternative which is called only when the Mono is empty.

### Actual behavior
switchIfEmpty takes a Mono.  As such, the Mono must be created even when it will never be subscribed to.

### Steps to reproduce
Given a class with these two functions:
```java
Mono<String> produceIt() {
... // Some code to produce the sought after result, asynchronously.
}
Mono<String> produceAlternative() {
... // Some code to produce the alternative asynchronously
}
```
Code in another member of that class:
```java
Mono<String> foo = produceIt().switchIfEmpty(produceAlternative());
```
as opposed to:
```java
Mono<String> foo = produceIt().switchIfEmpty(this::produceAlternative);
```

### Reactor Core version
3.1.5.RELEASE
### JVM version (e.g. `java -version`)
1.8.0_162 (Oracle Corporation 25.162-b12)

### Notes:
I will try to put together a PR for this sometime in the near future but I thought I would go ahead an submit it since it might be something really easy for you folks to add.

**Comment by OlegDokuka** [comment]
Hey @danapsimer. Thank you for that issue. What about simply having `produceIt().switchIfEmpty(Mono.defere(this::produceAlternative))` in that case?;

**Comment by danapsimer** [comment]
@OlegDokuka That would do the trick.  However I think it is more intuitive and consistent with other operations to provide the Supplier<Mono<?>> directly to the switchIfEmpty call.

**Comment by simonbasle** [comment]
I think it makes sense, and there's no ambiguity since there is no other `switchIfEmpty` overload that might take a lambda. @smaldini wdyt?

**Comment by simonbasle** [comment]
@danapsimer note however that often a `Flux<T>` by itself is lazy (a "cold" flux). That means that calling a method returning such a `Flux` to capture it in `switchIfEmpty` isn't that much of a problem. For "hot" flux, one that would trigger the processing immediately upon being instantiated, the `defer` wrapping suggested by @OlegDokuka is the standard way to go.

**Comment by OlegDokuka** [comment]
I will provide PR for that today

**Comment by davinkevin** [comment]
I could be interested by this overload. 

Especially because I sometimes have some code inside the `switchIfEmpty` which aren't reactive at core so it's triggering some actions even if don't need it.

**Comment by simonbasle** [comment]
@davinkevin As previously discussed `Mono.defer` is the way to go there rather than adding a `Supplier` overload, but that Kotlin extension is a good idea. . Would you like to contribute it to the `MonoExtensions.kt` and `FluxExtensions.kt` core extensions?

**Comment by davinkevin** [comment]
@simonbasle done in #1460 😉. I hope this will be accepted.

**Comment by ramez72** [comment]
Can anybody add this for java also? it really frustrating to add Mono.defer always.

**Comment by simonbasle** [comment]
@ramez72 nope, favor combination (`defer`) over covering all possible usecases via overloads.

**Comment by ramez72** [comment]
Ya, defer can do that, due to reactive nature my code adds lot of close Parentheses at the end, I thought it will reduce at least one (improved readability)
and beginners may not use defer (off-course I did while starting) which may result in unnecessary code execution and they won't figure out until there is a logging in the code which executed in switch part.
